### ISSUE TITLE
Sending and Receiving phases in mitmdump are very slow

### ISSUE THREAD

**Issue by frolino** [mitmproxy/mitmproxy#3802]
#### Problem Description
We're thinking of using mitmdump to capture network requests as part of our performance test (we were initially using Fiddler, but unfortunately, Fiddler does not support HTTP2). However, **after transitioning from Fiddler to mitmdump, we noticed that *all* the network requests have slowed down**. Digging deeper, it seems like the slowdown occurs in the Sending and Receiving phases:

![sending_receiving](https://user-images.githubusercontent.com/3180563/73500831-9e592400-4378-11ea-982e-ab9fa5e0aef5.PNG)
 
Notice in the image above how Sending takes almost 200ms, and Receiving takes almost 400ms (exact magnitude varies per request). On the other hand, **when using Fiddler, the Sending and Receiving phases are pretty much instant (i.e., ~0ms)**. This applies not only to requests sent by our web application - we can reproduce the behaviour using pretty much any other website (e.g., reddit).

Note that we tried to set the stream_large_bodies option to 1k, as we thought this might speed things up. However, the response times remained the same. A couple of questions:

1. Are there any other mitmdump settings/options that we can try to set, which may bring the response time closer to those from Fiddler's?
2. If no, do you know why the Sending and Receiving phases are taking so long? (perhaps it's some processing that mitmdump is doing that we don't actually require on our end, so we can just customize the mitm code for our purpose).

#### Steps to reproduce the behavior:
1. Start the mitmdump proxy. This is the command that we use (replace `<regex>` and `<filter>` with appropriate values):
    1. `mitmdump --ignore-hosts "<regex>" -k -w Requests.mitm "<filter>" `
2. Open the web application (any website would do - we recently tried it on www.reddit.com and we're able to reproduce the issue there)
3. Stop the mitmdump proxy, and save the requests captured by mitmdump (we converted ours to a .har file)
4. Start Fiddler
5. Open the same web application
6. Stop Fiddler, and save the requests captured by Fiddler
7. Compare the requests captured by mitmdump and those captured by Fiddler, and notice that the requests for the former are much slower (particularly the Sending and Receiving phases).

#### System Information
Mitmproxy: 4.0.4
Python:    3.7.0
OpenSSL:   OpenSSL 1.1.0i  14 Aug 2018
Platform:  Windows-10-10.0.14393-SP0

**Comment by mhils** [comment]
> Are there any other mitmdump settings/options that we can try to set, which may bring the response time closer to those from Fiddler's?

No.

> If no, do you know why the Sending and Receiving phases are taking so long? (perhaps it's some processing that mitmdump is doing that we don't actually require on our end, so we can just customize the mitm code for our purpose).

Frankly speaking, performance optimizations haven't been much of a concern for us so far as it's simply not required for most use cases. If you want to look into this:

1. This may be OS-specific, due to how sockets/threading works on Python on Windows.
2. It may be useful to do WireShark captures + log timestamps for our addon events (https://docs.mitmproxy.org/stable/addons-events/) and see where the majority of time is spent.

I'd be eager to hear if you find anything interesting!

**Comment by frolino** [comment]
Thanks for the quick reply - will try those out. Btw, we recently tried setting stream_large_bodies to 1 (i.e., stream all requests) while running mitmdump on two different Windows machines, and we observed two different behaviours. Streaming had a noticeable order of magnitude impact on the "Receive" time in one machine (went from 200-300ms to 0-50ms), but had practically no impact on the other. Unfortunately, it's the latter machine that we need to run our tests on. It's possible that this could be because of machine configuration differences, but if you think there's anything else obvious we might be able to look into, let me know :)

Also, I noticed that the "Send" time wasn't impacted at all after switching to streaming. Does stream_large_bodies only apply to the response, and not the request?
